### PR TITLE
docstrings: fixes, also point to the new guide chapter

### PR DESCRIPTION
**Pull request by birkenfeld**
I hope the links are okay like this?

### CODE REVIEW

File: src/instance.rs [hunk 1]
```diff
@@ -13,19 +13,27 @@ use std::marker::PhantomData;
...
 }
 
-/// Safe wrapper around unsafe `*mut ffi::PyObject` pointer with specified type information.
+/// A Python object of known type.
 ///
-/// `Py<T>` is thread-safe, because any python related operations require a Python<'p> token.
+/// Accessing this object is thread-safe, since any access to its API requires a
+/// `Python<'py>` GIL token.
+///
+/// See [the guide](https://pyo3.rs/v0.9.0-alpha.1/types.html) for an explanation
```
Comment (kngwyu) [inline src/instance.rs]:
```suggestion
/// See [the guide](https://pyo3.rs/master/types.html) for an explanation
```

File: src/lib.rs [hunk 2]
```diff
@@ -7,23 +7,26 @@
...
-//! You obtain a `Python` instance by acquiring the GIL,
-//! and have to pass it into all operations that call into the Python runtime.
+//! Because all Python objects potentially have multiple owners, the concept of
+//! Rust mutability does not apply to Python objects.  As a result, PyO3 allows
+//! mutating Python objects even if they are not stored in a mutable Rust
+//! variable.
+//!
+//! In Python, all objects are implicitly reference counted.  The Python
+//! interpreter uses a global interpreter lock (GIL) to ensure thread-safety.
+//! This API uses a zero-sized `struct Python<'py>` as a token to indicate that
```
Comment (kngwyu) [inline src/lib.rs]:
```suggestion
//! Thus, we use `struct Python<'py>` as a token to indicate that
```
The old `This API` refers to `PyObject` but the sentence it occurs is removed now.

File: src/lib.rs [hunk 3]
```diff
@@ -84,13 +91,21 @@
...
+//! [maturin](https://github.com/PyO3/maturin) or
+//! [setuptools-rust](https://github.com/PyO3/setuptools-rust). You can find an
+//! example for setuptools-rust in [examples/word-count](examples/word-count),
+//! while maturin should work on your crate without any configuration.
 //!
-//! ## Using python from rust
+//! ## Using Python from Rust
 //!
-//! Add `pyo3` this to your `Cargo.toml`:
+//! Add `pyo3` to your `Cargo.toml`:
```
Comment (kngwyu) [inline src/lib.rs]: :+1:

File: src/types/any.rs [hunk 4]
```diff
@@ -3,12 +3,18 @@ use crate::err::PyDowncastError;
...
+/// A Python object with GIL lifetime
 ///
+/// Represents any Python object.  All Python objects can be cast to `PyAny`.
 /// In addition, if the inner object is an instance of type `T`, we can downcast
 /// `PyAny` into `T`.
 ///
+/// `PyAny` is used as a reference with a lifetime that represents that the GIL
+/// is held, therefore its API does not require a `Python<'py>` token.
+///
+/// See [the guide](https://pyo3.rs/v0.9.0-alpha.1/types.html) for an explanation
```
Comment (kngwyu) [inline src/types/any.rs]:
```suggestion
/// See [the guide](https://pyo3.rs/master/types.html) for an explanation
```

File: src/object.rs [hunk 5]
```diff
@@ -9,10 +9,13 @@ use crate::{AsPyPointer, Py, Python};
...
 
-/// A python object
+/// A Python object of any type.
 ///
-/// The python object's lifetime is managed by python's garbage
-/// collector.
+/// The Python object's lifetime is managed by Python's garbage collector,
+/// so to access the object API, a `Python<'py>` GIL token is required.
+///
+/// See [the guide](https://pyo3.rs/v0.9.0-alpha.1/types.html) for an explanation
```
Comment (kngwyu) [inline src/object.rs]:
```suggestion
/// See [the guide](https://pyo3.rs/master/types.html) for an explanation
```

File: src/lib.rs (hunk 2 above)
Comment (birkenfeld) [reply src/lib.rs]: Indeed, I glued this in too eagerly :)

File: src/lib.rs [hunk 6]
```diff
@@ -7,23 +7,26 @@
...
+//! mutating Python objects even if they are not stored in a mutable Rust
+//! variable.
+//!
+//! In Python, all objects are implicitly reference counted.  The Python
+//! interpreter uses a global interpreter lock (GIL) to ensure thread-safety.
+//! Thus, we use `struct Python<'py>` as a token to indicate that
+//! a function can assume that the GIL is held.  In Rust, we use different types
+//! to represent a reference to a Python object, depending on whether we know
+//! the GIL is held, and depending on whether we know the underlying type.  See
+//! [the guide](https://pyo3.rs/v0.9.0-alpha.1/types.html) for an explanation of
```
Comment (birkenfeld) [inline src/lib.rs]:
```suggestion
//! [the guide](https://pyo3.rs/master/types.html) for an explanation of
```

**Comment by birkenfeld** [comment]
OK, "small" update, I also made a lot of consistency fixes to the individual docstrings together with the occasional real update.

File: src/python.rs [hunk 7]
```diff
@@ -105,19 +106,26 @@ impl<'p> Python<'p> {
...
     /// ```
     ///
-    /// **NOTE**
-    /// You cannot use all `&Py~` types in the closure that `allow_threads` takes.
+    /// **Note:**
+    /// PyO3 types that represent objects with a lifetime tied to holding the GIL
+    /// cannot be used in the closure.  This includes `&PyAny` and all the
+    /// concrete-typed siblings, like `&PyString`.
+    ///
+    /// You can convert such references to e.g. `PyObject` or `Py<PyString>`,
```
Comment (kngwyu) [inline src/python.rs]: I really like this explanation :+1:

File: src/err.rs [hunk 8]
```diff
@@ -69,13 +69,22 @@ impl PyErr {
...
-    /// * any other value: the exception instance will be created using python `T(value)`
+    /// * a tuple: the exception instance will be created using Python `T(*tuple)`
+    /// * any other value: the exception instance will be created using Python `T(value)`
     ///
-    /// Panics if `T` is not a python class derived from `BaseException`.
+    /// Panics if `T` is not a Python class derived from `BaseException`.
     ///
     /// Example:
-    ///  `return Err(PyErr::new::<exceptions::TypeError, _>("Error message"));`
+    /// ```
```
Comment (kngwyu) [inline src/err.rs]:
```suggestion
    /// ```ignore
```

File: src/err.rs [hunk 9]
```diff
@@ -69,13 +69,22 @@ impl PyErr {
...
+    /// return Err(PyErr::new::<exceptions::TypeError, _>("Error message"));
+    /// ```
+    ///
+    /// In most cases, you can use a concrete exception's constructors instead:
+    /// the example is equivalent to
+    /// ```
```
Comment (kngwyu) [inline src/err.rs]:
```suggestion
    /// ```ignore
```

File: src/instance.rs [hunk 10]
```diff
@@ -13,19 +13,27 @@ use std::marker::PhantomData;
 use std::mem;
 use std::ptr::NonNull;
 
-/// Types that are built into the python interpreter.
+/// Types that are built into the Python interpreter.
 ///
-/// pyo3 is designed in a way that that all references to those types are bound to the GIL,
-/// which is why you can get a token from all references of those types.
+/// PyO3 is designed in a way that that all references to those types are bound
```
Comment (Alexander-N) [inline src/instance.rs]:
```suggestion
/// PyO3 is designed in a way that all references to those types are bound
```

File: src/class/methods.rs [hunk 11]
```diff
@@ -133,7 +133,7 @@ pub trait PyMethodsInventory: inventory::Collect {
     fn get_methods(&self) -> &'static [PyMethodDefType];
 }
 
-/// The implementation of tis trait defines which methods a python type has.
+/// The implementation of tis trait defines which methods a Python type has.
```
Comment (Alexander-N) [inline src/class/methods.rs]:
```suggestion
/// The implementation of this trait defines which methods a Python type has.
```

**Pull request merged**